pkg/providers/cloudstack/validator.go

```diff
@@ -0,0 +1,197 @@
+		workerNodeGroupMachineConfig.Spec.Users[0].SshAuthorizedKeys = []string{""}
+	}
+
+	err := v.validateControlPlaneHost(&clusterSpec.Cluster.Spec.ControlPlaneConfiguration.Endpoint.Host)
+	if err != nil {
+		return err
+	}
+
+	for _, machineConfig := range v.machineConfigs {
+		if machineConfig.Namespace != clusterSpec.Namespace {
+			return errors.New("CloudStackMachineConfig and Cluster objects must have the same namespace specified")
+		}
+	}
+	if clusterSpec.datacenterConfig.Namespace != clusterSpec.Namespace {
+		return errors.New("CloudStackDatacenterConfig and Cluster objects must have the same namespace specified")
+	}
+
+	if controlPlaneMachineConfig.Spec.Template.Value == "" {
+		return fmt.Errorf("control plane CloudStackMachineConfig template is not set. Default template is not supported in CloudStack, please provide a template name")
+	}
+
+	if err = v.validateMachineConfig(ctx, clusterSpec.datacenterConfig.Spec, controlPlaneMachineConfig); err != nil {
+		logger.V(1).Info("Control plane machine config validation failed.")
+		return err
+	}
+	if controlPlaneMachineConfig.Spec.Template != workerNodeGroupMachineConfig.Spec.Template {
+		if workerNodeGroupMachineConfig.Spec.Template.Value == "" {
+			return fmt.Errorf("worker CloudStackMachineConfig template is not set. Default template is not supported in CloudStack, please provide a template name")
+		}
+		if err = v.validateMachineConfig(ctx, clusterSpec.datacenterConfig.Spec, workerNodeGroupMachineConfig); err != nil {
+			logger.V(1).Info("Workload machine config validation failed.")
+			return err
+		}
+		if controlPlaneMachineConfig.Spec.Template != workerNodeGroupMachineConfig.Spec.Template {
+			return errors.New("control plane and worker nodes must have the same template specified")
+		}
+	}
+	logger.MarkPass("Control plane and Workload templates validated")
+
+	if etcdMachineConfig != nil {
+		if etcdMachineConfig.Spec.Template.Value == "" {
+			return fmt.Errorf("etcd CloudStackMachineConfig template is not set. Default template is not supported in CloudStack, please provide a template name")
+		}
+		if err = v.validateMachineConfig(ctx, clusterSpec.datacenterConfig.Spec, etcdMachineConfig); err != nil {
+			logger.V(1).Info("Etcd machine config validation failed.")
+			return err
+		}
+		if etcdMachineConfig.Spec.Template != controlPlaneMachineConfig.Spec.Template {
+			return errors.New("control plane and etcd machines must have the same template specified")
+		}
+	}
+
+	return nil
+}
+
+func (v *Validator) validateMachineConfig(ctx context.Context, datacenterConfigSpec anywherev1.CloudStackDatacenterConfigSpec, machineConfig *anywherev1.CloudStackMachineConfig) error {
+	domain := datacenterConfigSpec.Domain
+	zone := datacenterConfigSpec.Zone
+	account := datacenterConfigSpec.Account
+	if err := v.cmk.ValidateTemplatePresent(ctx, domain, zone, account, machineConfig.Spec.Template); err != nil {
+		return fmt.Errorf("error validating template: %v", err)
```

Comment:
As a general comment, applying the word `error` when adding context is superfluous. We know its an error, that's how its reported to the user. The context need only shed light on the circumstances: `validating template: %v`.

I know we use the word `error` in lots of places, but think about how that presents: `error running create: error validating template: error with foo: error with bar`. It becomes a little obnoxious 😄

Reply:
That's a good point that we didn't really consider before, I would say `unable to validate template` or something is fine too, a bit more descriptive of the action. I think there is value in bubbling up with some context as to what kind of error validating template is, because it could make it seem like there was an error validating after finding the template or something.